### PR TITLE
NEW: Allow search field customisation

### PR DESCRIPTION
**Pull request by unclecheese**
By default, `GridFieldFilterHeader` chooses the first field in the searchable fields array for the search bar. This allows you to customise what that field is.

## Notes
- Assumes https://github.com/silverstripe/silverstripe-framework/pull/10382 will be merged in first.
- The reviewer should make a judgment call as to whether this is a useful enhancement following #10382. I (Guy) think it _probably_ is, as it allows people to customise their `SearchContext` and `GridField`s further and separately.

## Parent issue
- https://github.com/silverstripe/silverstripe-framework/issues/9356

### CODE REVIEW

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (maxime-rainville) [inline src/Forms/GridField/GridFieldFilterHeader.php]: I would make this private with the expectation that child class that need to interact with it should use/override the accessor/mutator.

```suggestion
    private $searchField = null;
```

Comment (unclecheese) [reply src/Forms/GridField/GridFieldFilterHeader.php]: My gut feeling was private, but the rest of the api isn't private, so it feels awkward to me. But I'll defer to you here.

**Comment by unclecheese** [comment]
I'm not totally sold on adding it to the `DataObject` API. I think we've already got such a bloated  `DataObject` "config", and we have a long history of just chucking concerns on to that pile that could be handled elsewhere.

**Comment by maxime-rainville** [comment]
That's probably a wider discussion for an RFC. This PR by itself doesn't preclude our duplicate these efforts. I'm fine with merging this PR once unit tests have been added.

**Comment by unclecheese** [comment]
Good feedback, @maxime-rainville .. have added the `primary_search_field` config, plus tests and docs.

**Comment by Cheddam** [comment]
Likely doesn't impact this PR going through as-is, but #9356 talks about setting _multiple_ fields to be searchable in the main field. Thinking we could change these new APIs in the future to optionally accept an array to achieve that issue's goals?

**Comment by lekoala** [comment]
@Cheddam thanks for tracking this. Multiple field is indeed necessary in my opinion, as it is very unconvenient for the user to go through the dropdown in order to search on other fields. I think the common expectation of all my customers is that "typing something in the search bar matches on all relevant fields", not only on a single field.

**Comment by Cheddam** [comment]
@lekoala I appreciate that this feature would be really useful to users, but it'll need to be tackled in a separate PR - I just wanted to make @unclecheese aware of it, and check that this API is going to suit expansion to support it in the future.

File: docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md
Comment (maxime-rainville) [inline docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md]:
```suggestion
Tabular views such as `GridField` or `ModalAdmin` include a search bar. By default, the search bar will search against the *first
```

**Review by maxime-rainville — DISMISSED** [review]
I'm happy with the PR and I've re-tested that it works. I'm thinking some of the doc could use simpler language, but otherwise I'm happy.

**Comment by emteknetnz** [comment]
@unclecheese when you get a chance would you be able to implement Maxime's recommended docs changes? We should be able to get this merged once those are in

**Comment by lekoala** [comment]
For reference
https://github.com/silverstripe/silverstripe-framework/pull/9836

**Comment by GuySartorelli** [comment]
I have rebased this on `4`, resolved merge conflicts, and updated the docs.
The documentation assumes https://github.com/silverstripe/silverstripe-framework/pull/10382 will be merged before this. There will likely be a merge conflict between these two PRs' docs, as well.

**Comment by GuySartorelli** [comment]
Note: The documentation portion should not be merged here. I have started [the process of migrating docs](https://github.com/silverstripe/silverstripe-framework/issues/10215) to [a new repository](https://github.com/silverstripe/doc) so the docs will need to be removed from this PR and a new PR raised in the new repo instead.

**Comment by GuySartorelli** [comment]
Docs changes have been moved to https://github.com/silverstripe/developer-docs/pull/6

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (emteknetnz) [inline src/Forms/GridField/GridFieldFilterHeader.php:136]:
```suggestion
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Could you move the contents of this PR into https://github.com/silverstripe/silverstripe-framework/pull/10382 and close this one, that way there's only a single framework PR which makes testing a lot easier

**Comment by GuySartorelli** [comment]
If you agree the changes in this PR are worthwhile keeping then yes, I'll merge the two PRs. Though the combined PR should not be squashed in that case. These are two related but separate changes.

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldFilterHeader.php]: done

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldFilterHeader.php]: done

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldFilterHeader.php]: done

File: src/Forms/GridField/GridFieldFilterHeader.php
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldFilterHeader.php:142]: Subjective but okay. Done.

**Comment by emteknetnz** [comment]
Oh right, sorry got confused with all the PRs. Hold off for now, see how the other PR looks on it's own and then in a better position to judge if this is worthwhile

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I'm thinking close this one and add an allowlist style config into https://github.com/silverstripe/silverstripe-framework/pull/10382 would be better

**Comment by GuySartorelli** [comment]
We came to the decision not to use an allow-list configuration in the end: https://github.com/silverstripe/silverstripe-framework/pull/10382#issuecomment-1181240324

**Review by emteknetnz — APPROVED** [review]
Tested locally, works good

**Pull request merged**
